File: src/main/javascript/crypto/src/libs/AttestedObject.ts
import {AttestationCrypto} from "./AttestationCrypto";
import {SignedIdentifierAttestation} from "./SignedIdentifierAttestation";
import {hexStringToArray, logger, uint8tohex} from "./utils";
import {Asn1Der} from "./DerUtility";
import {ProofOfExponentInterface} from "./ProofOfExponentInterface";
import {KeyPair, KeysArray} from "./KeyPair";
import {AsnParser} from "@peculiar/asn1-schema";
import {UseToken} from "../asn1/shemas/UseToken";
import {UsageProofOfExponent} from "./UsageProofOfExponent";
import {IdentifierAttestation} from "./IdentifierAttestation";
import {Attestable} from "./Attestable";
import {Verifiable} from "./Verifiable";
import {ASNEncodable} from "./ASNEncodable";
import {AttestableObject} from "./AttestableObject";
import {DEBUGLEVEL} from "../config";

declare global {
    interface Window {
        ethereum: any;
        web3: any;
    }
}


export class AttestedObject implements ASNEncodable, Verifiable {
    private crypto: AttestationCrypto;
    private pok: ProofOfExponentInterface;
    private derEncodedProof: string;
    private encoding: string;
    private attestableObject: any;
    private att: SignedIdentifierAttestation;
    private attestationSecret: bigint;
    private objectSecret: bigint;
    private userKeyPair: KeyPair;

    private preSignEncoded: string;

    private webDomain: string;

    static Eip712UserData: {[index: string]:string|number}  = {
        payload: '',
        description: '',
        timestamp: 0
    }
    // static Eip712UserDataTypes: {[index: string]:string}[]  = [
    static Eip712UserDataTypes: {name: string, type: string}[]  = [
        {name: 'payload', type: 'string'},
        {name: 'description', type: 'string'},
        {name: 'timestamp', type: 'uint256'},
    ]
    static Eip712UserDataPrimaryName: string = "Authentication";
    static Eip712UserDataDescription: string = "Single-use authentication";

    constructor() {}

    create<T extends Attestable>(
        attestableObject: T ,
        att: SignedIdentifierAttestation,
        attestationSecret: bigint ,
        objectSecret: bigint
    ){
        this.attestableObject = attestableObject;
        this.att = att;
        this.attestationSecret = attestationSecret;
        this.objectSecret = objectSecret;
        this.crypto = new AttestationCrypto();
        this.pok = this.makeProof(attestationSecret, objectSecret, this.crypto);
        this.derEncodedProof = this.pok.getDerEncoding();

        this.fillPresignData();

    }

    setWebDomain(domain: string){
        this.webDomain = domain;
    }

    fillPresignData(){
        this.preSignEncoded = 
            this.attestableObject.getDerEncoding() +
            this.att.getDerEncoding() +
            this.pok.getDerEncoding();

        this.encoding = Asn1Der.encode('SEQUENCE_30', this.preSignEncoded);
    }

    fromDecodedData<T extends Attestable>(
        attestableObject: T ,
        att: SignedIdentifierAttestation,
        pok: ProofOfExponentInterface
    ){
        this.attestableObject = attestableObject;
        this.att = att;
        this.pok = pok;

        this.fillPresignData();

        this.userKeyPair = this.att.getUnsignedAttestation().getSubjectPublicKeyInfo();

        this.constructorCheck();
    }


    // async sign(){
    //     let userData = {
    //         payload: this.encoding,
    //         description: AttestedObject.Eip712UserDataDescription,
    //         timestamp: new Date().getTime()
    //     };
    //
    //     return await SignatureUtility.signEIP712WithBrowserWallet(this.webDomain, userData, AttestedObject.Eip712UserDataTypes, AttestedObject.Eip712UserDataPrimaryName );
    // }

    public verify(): boolean{
        if (!this.attestableObject.verify()) {
            logger(DEBUGLEVEL.LOW, "Could not verify attestable object");
            return false;
        }
        if (!this.att.verify()) {
            logger(DEBUGLEVEL.LOW, "Could not verify attestation");
            return false;
        }
        if (!this.crypto.verifyEqualityProof(
            this.att.getUnsignedAttestation().getCommitment(),
            this.attestableObject.getCommitment(),
            this.pok
        )) {
            logger(DEBUGLEVEL.LOW, "Could not verify the consistency between the commitment in the attestation and the attested object");
            return false;
        }

        return true;
    }

    static fromBytes<D extends UseToken, T extends AttestableObject>(asn1: Uint8Array, decoder: new () => D, attestorKey: KeyPair, attestable: new () => T, issuerKeys: KeysArray): AttestedObject{

        let attested: D = AsnParser.parse(asn1, decoder);

        let me = new this();

        me.attestableObject = new attestable();
        me.attestableObject.fromBytes(attested.signedToken, issuerKeys);

        me.att = SignedIdentifierAttestation.fromBytes(new Uint8Array(attested.attestation), attestorKey);

        let pok = new UsageProofOfExponent();
        pok.fromBytes( new Uint8Array(attested.proof) ) ;
        me.pok = pok;

        me.userKeyPair = me.att.getUnsignedAttestation().getSubjectPublicKeyInfo();

        // TODO: Use getter to instantiate AttestationCrypto when required
        me.crypto = new AttestationCrypto();

        me.constructorCheck();

        return me;
    }

    private makeProof(attestationSecret: bigint, objectSecret: bigint, crypto: AttestationCrypto): ProofOfExponentInterface {
        // TODO Bob should actually verify the attestable object is valid before trying to cash it to avoid wasting gas
        // Need to decode twice since the standard ASN1 encodes the octet string in an octet string
        // TODO we dont parse that value, because its already parsed to this.riddle
        // let extensions = this.att.getUnsignedAttestation().getExtensions();//.getObjectAt(0));

        // Index in the second DER sequence is 2 since the third object in an extension is the actual value

        // TODO we dont parse that value, because its already parsed to this.riddle
        // let attCom: Uint8Array = new Uint8Array(extensions.extension.extnValue);
        let attCom: Uint8Array = this.att.getUnsignedAttestation().getCommitment();
        let objCom: Uint8Array = this.attestableObject.getCommitment();
        let pok: ProofOfExponentInterface = crypto.computeEqualityProof(uint8tohex(attCom), uint8tohex(objCom), attestationSecret, objectSecret);

        if (!crypto.verifyEqualityProof(attCom, objCom, pok)) {
            throw new Error("The redeem proof did not verify");
        }
        return pok;
    }

    getAttestableObject(){
        return this.attestableObject;
    }

    getAtt(){
        return this.att;
    }

    getDerEncodeProof(){
        return this.derEncodedProof;
    }

    public getDerEncoding():string {
        return this.encoding;
    }

    private constructorCheck() {
        if (!this.verify()) {
            throw new Error("The redeem request is not valid");
        }
    }

    public checkValidity(ethAddress:string = ""): boolean {

        // CHECK: that it is an identifier attestation otherwise not all the checks of validity needed gets carried out
        try {
            let attEncoded = this.att.getUnsignedAttestation().getDerEncoding();
            let std: IdentifierAttestation = IdentifierAttestation.fromBytes(new Uint8Array(hexStringToArray(attEncoded))) as IdentifierAttestation;

            // CHECK: perform the needed checks of an identifier attestation
            if (!std.checkValidity()) {
                logger(DEBUGLEVEL.LOW, "The attestation is not a valid standard attestation");
                return false;
            }
        } catch (e) {
            logger(DEBUGLEVEL.LOW, "The attestation is invalid");
            return false;
        }

        try {
            // CHECK: that the cheque is still valid
            if (!this.getAttestableObject().checkValidity()) {
                logger(DEBUGLEVEL.LOW, "Cheque is not valid");
                return false;
            }
        } catch (e) {
            logger(DEBUGLEVEL.LOW, "Cheque validation failed");
            return false;
        }

		if (!ethAddress)
			return true;

        try {

            // CHECK: the Ethereum address on the attestation matches receivers signing key
            // let attestationEthereumAddress: string = this.getAtt().getUnsignedAttestation().getSubject().substring(3);
            let attestationEthereumAddress: string = this.getAtt().getUnsignedAttestation().getAddress();
            logger(DEBUGLEVEL.HIGH, 'attestationEthereumAddress: ' + attestationEthereumAddress);
            logger(DEBUGLEVEL.HIGH, 'providedEthereumAddress: ' + ethAddress);

            if (attestationEthereumAddress.toLowerCase() !== ethAddress.toLowerCase()) {
                logger(DEBUGLEVEL.LOW, "The attestation is not to the same Ethereum user who is sending this request");
                return false;
            }
        } catch (e) {
            logger(DEBUGLEVEL.LOW, "Address validation failed");
            logger(DEBUGLEVEL.MEDIUM, e);
            return false;
        }

        return true;
    }
}

File: src/main/java/org/tokenscript/attestation/AttestedObject.java
package org.tokenscript.attestation;

import java.io.IOException;
import java.io.InvalidObjectException;
import java.math.BigInteger;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.asn1.ASN1EncodableVector;
import org.bouncycastle.asn1.ASN1InputStream;
import org.bouncycastle.asn1.ASN1Sequence;
import org.bouncycastle.asn1.DERSequence;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.util.PublicKeyFactory;
import org.tokenscript.attestation.core.Attestable;
import org.tokenscript.attestation.core.AttestationCrypto;
import org.tokenscript.attestation.core.ExceptionUtil;
import org.tokenscript.attestation.core.SignatureUtility;

public class AttestedObject<T extends Attestable> extends AttestedKeyObject {
  private static final Logger logger = LogManager.getLogger(AttestedObject.class);
  private final T attestableObject;
  private final SignedIdentifierAttestation att;
  private final ProofOfExponent pok;

  private final AsymmetricKeyParameter userPublicKey;

  private final byte[] encoding;

  public AttestedObject(T attestableObject, SignedIdentifierAttestation att, AsymmetricKeyParameter userPublicKey,
      BigInteger attestationSecret, BigInteger chequeSecret, AttestationCrypto crypto) {
    this(attestableObject, att, attestationSecret, chequeSecret, new byte[0], crypto);
  }

  public AttestedObject(T attestableObject, SignedIdentifierAttestation att,
      BigInteger attestationSecret, BigInteger chequeSecret, byte[] unpredictableNumber,
      AttestationCrypto crypto)
  {
    this.attestableObject = attestableObject;
    this.att = att;

    try {
      this.userPublicKey = PublicKeyFactory.createKey(att.getUnsignedAttestation().getSubjectPublicKeyInfo());
      this.pok = makeProof(attestationSecret, chequeSecret, unpredictableNumber, crypto);
      ASN1EncodableVector vec = new ASN1EncodableVector();
      vec.add(ASN1Sequence.getInstance(this.attestableObject.getDerEncoding()));
      vec.add(ASN1Sequence.getInstance(att.getDerEncoding()));
      vec.add(ASN1Sequence.getInstance(pok.getDerEncoding()));
      this.encoding = new DERSequence(vec).getEncoded();
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not decode asn1", e);
    }
    constructorCheck();
  }

  public AttestedObject(T object, SignedIdentifierAttestation att, ProofOfExponent pok) {
    this.attestableObject = object;
    this.att = att;
    this.pok = pok;

    try {
      ASN1EncodableVector vec = new ASN1EncodableVector();
      vec.add(ASN1Sequence.getInstance(object.getDerEncoding()));
      vec.add(ASN1Sequence.getInstance(att.getDerEncoding()));
      vec.add(ASN1Sequence.getInstance(pok.getDerEncoding()));
      this.encoding = new DERSequence(vec).getEncoded();
      this.userPublicKey = PublicKeyFactory.createKey(att.getUnsignedAttestation().getSubjectPublicKeyInfo());
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not encode asn1", e);
    }
    constructorCheck();
  }

  public AttestedObject(byte[] derEncoding, ObjectDecoder<T> attestableObjectDecoder,
                        AsymmetricKeyParameter publicAttestationSigningKey) {
    this.encoding = derEncoding;
    try {
      ASN1InputStream input = new ASN1InputStream(derEncoding);
      ASN1Sequence asn1 = ASN1Sequence.getInstance(input.readObject());
      input.close();
      this.attestableObject = attestableObjectDecoder.decode(asn1.getObjectAt(0).toASN1Primitive().getEncoded());
      this.att = new SignedIdentifierAttestation(asn1.getObjectAt(1).toASN1Primitive().getEncoded(), publicAttestationSigningKey);
      this.pok = new UsageProofOfExponent(asn1.getObjectAt(2).toASN1Primitive().getEncoded());
      this.userPublicKey = PublicKeyFactory.createKey(att.getUnsignedAttestation().getSubjectPublicKeyInfo());
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not decode asn1", e);
    }
    constructorCheck();
  }

  private void constructorCheck() {
    if (!verify()) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Could not verify object"));
    }
  }

  public T getAttestableObject() {
    return attestableObject;
  }

  public SignedIdentifierAttestation getAtt() {
    return att;
  }

  public ProofOfExponent getPok() {
    return pok;
  }

  @Override
  public AsymmetricKeyParameter getAttestedUserKey() {
    return userPublicKey;
  }

  /**
   * Verifies that the redeem request will be accepted by the smart contract
   * @return true if the redeem request should be accepted by the smart contract
   */
  @Override
  public boolean checkValidity() {
    // CHECK: that it is an identifier attestation otherwise not all the checks of validity needed gets carried out
    try {
      byte[] attEncoded = att.getUnsignedAttestation().getDerEncoding();
      IdentifierAttestation std = new IdentifierAttestation(attEncoded);
      // CHECK: perform the needed checks of an identifier attestation
      if (!std.checkValidity()) {
        logger.error("The attestation is not a valid standard attestation");
        return false;
      }
    } catch (InvalidObjectException e) {
      logger.error("The attestation is invalid");
      return false;
    } catch (IOException e) {
      logger.error("The attestation could not be parsed as a standard attestation");
      return false;
    }

    // CHECK: that the object is still valid
    if (!getAttestableObject().checkValidity()) {
      logger.error("Object is not valid");
      return false;
    }

    // CHECK: the Ethereum address on the attestation matches receivers signing key
    String attestationEthereumAddress = getAtt().getUnsignedAttestation().getAddress();
    if (!attestationEthereumAddress.equals(SignatureUtility.addressFromKey(getAttestedUserKey()))) {
      logger.error("The attestation is not to the same Ethereum user who is sending this request");
      return false;
    }
    return true;
  }

  @Override
  public boolean verify() {
    if (!attestableObject.verify()) {
      logger.error("Could not verify attestable object");
      return false;
    }
    if (!att.verify()) {
      logger.error("Could not verify attestation");
      return false;
    }
    if (!AttestationCrypto.verifyEqualityProof(att.getUnsignedAttestation().getCommitment(), attestableObject.getCommitment(), pok)) {
      logger.error("Could not verify the consistency between the commitment in the attestation and the attested object");
      return false;
    }
    return true;
  }

  private UsageProofOfExponent makeProof(BigInteger attestationSecret, BigInteger objectSecret, byte[] unpredictableNumber, AttestationCrypto crypto) {
    // TODO Bob should actually verify the attestable object is valid before trying to work with on the blockchain it to avoid wasting gas
    // We require that the internal attestation is an IdentifierAttestation
    UsageProofOfExponent pok = crypto.computeEqualityProof(att.getUnsignedAttestation().getCommitment(), attestableObject.getCommitment(), attestationSecret, objectSecret, unpredictableNumber);
    if (!crypto.verifyEqualityProof(att.getUnsignedAttestation().getCommitment(), attestableObject.getCommitment(), pok)) {
      throw ExceptionUtil.throwException(logger,
          new RuntimeException("The redeem proof did not verify"));
    }
    return pok;
  }

  @Override
  public byte[] getDerEncoding() {
    return encoding;
  }

  // TODO override equals and hashcode
}
